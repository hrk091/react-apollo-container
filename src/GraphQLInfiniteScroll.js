// @flow

import React, { Component } from "react"
import InfiniteScroll from "react-infinite-scroller"
import type { ConnectionQueryData, ConnectionQueryProps } from "./types"


export function graphqlConcatResultNodes<A, T>(
  typename: A,
  prev: ConnectionQueryData<A, T>,
  next: { fetchMoreResult: ConnectionQueryData<A, T> }
): ConnectionQueryData<A, T> {
  if (!next || !next.fetchMoreResult) {
    console.warn(`fetchMore result might be wrong: ${next}`)
    return prev
  }
  const { fetchMoreResult } = next
  const newEdges = fetchMoreResult[typename].edges
  const pageInfo = fetchMoreResult[typename].pageInfo
  let res: ConnectionQueryData<A, T>
  // fetchMore中に別のqueryが実行されるとタイミングによってprevがundefinedになりうる
  if (prev) {
    res = {
      // $FlowFixMe: typenameはgenericsでinjectionするが、型はstring
      [typename]: Object.assign({}, prev[typename], {
        edges: [...prev[typename].edges, ...newEdges],
        pageInfo
      })
    }
  } else {
    res = {
      // $FlowFixMe
      [typename]: Object.assign(
        {},
        {
          edges: [...newEdges],
          pageInfo,
          // 下記がないとwarningが出る
          totalCount: 0,
          __typename: ""
        }
      )
    }
  }
  return res
}

export default class GraphQLInfiniteScroll<A, T> extends Component<ConnectionQueryProps<A, T>> {
  /**
   * graphQLと連携し、infinite scroll paginationを提供するcomponent
   * Relay formatのpagenationに対応したgraphQLに対応
  */
  onLoadMore = () => {
    const { typename, data } = this.props
    if (!data || !data[typename] || !data[typename].pageInfo) {
      return
    }
    this.props.fetchMore({
      variables: {
        cursor: data[typename].pageInfo.endCursor
      },
      updateQuery: graphqlConcatResultNodes.bind(null, typename)
    })
  }

  render() {
    const { typename, data, loading, error } = this.props
    if (error) {
      console.error(error)
      return (
        <>
          <h3>データ取得エラー</h3>
          <h5>エラーが発生しました。ページをリロードしてください。</h5>
          <h5>事象が改善しない場合は、管理者にお問い合わせください。</h5>
        </>
      )
    }
    if (!data || !data[typename] || !data[typename].pageInfo) {
      return null
    }
    if (loading) {
      return <h3>Loading...</h3>
    }

    const { hasNextPage } = data[typename].pageInfo
    const records = data[typename].edges.filter(Boolean).map((edge: any) => edge.node)
    const { children: Children } = this.props

    return (
      <InfiniteScroll loadMore={this.onLoadMore} hasMore={hasNextPage} loader={<h3>Loading...</h3>} >
        <Children records={records} total={data[typename].totalCount} />
      </InfiniteScroll>
    )
  }
}
