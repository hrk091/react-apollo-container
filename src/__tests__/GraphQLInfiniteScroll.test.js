import React from "react"
import { MockedProvider } from "@apollo/react-testing"
import { render } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"

import { useQuery } from "@apollo/react-hooks"
import gql from "graphql-tag"

import { waitForDomChange } from "@testing-library/dom"
import GraphQLInfiniteScroll, { graphqlConcatResultNodes } from "../GraphQLInfiniteScroll"

describe("GraphQLInfiniteScroll", () => {
  // Setup

  const mockQuery = gql`
      query ContactList($cursor: String, $name: String) {
          allContact(first: 10, after: $cursor, name: $name) {
              totalCount
              pageInfo {
                  endCursor
                  hasNextPage
              }
              edges {
                  node {
                      id
                  }
              }
          }
      }
  `
  const variables = {
    name: "dummyName"
  }
  const typename = "allContact"

  function DummyComponent() {
    const queryProps = useQuery(mockQuery, {variables})
    return (
      <GraphQLInfiniteScroll typename={typename} {...queryProps}>
        {({records, total}) => {
          return <div/>
        }}
      </GraphQLInfiniteScroll>
    )
  }

  // Tests

  it("renders loading state initially", () => {
    // Given

    const mocks = [
      {
        request: {query: mockQuery, variables},
        result: {
          data: {
            [typename]: {
              totalCount: 0,
              pageInfo: {
                endCursor: Symbol("endCursor"),
                hasNextPage: false
              },
              edges: []
            }
          }
        }
      }
    ]

    const spy = jest.fn()

    function DummyComponent() {
      const queryProps = useQuery(mockQuery, {variables})
      return (
        <GraphQLInfiniteScroll typename={typename} {...queryProps}>
          {({records, total}) => {
            spy()
            return <div/>
          }}
        </GraphQLInfiniteScroll>
      )
    }

    // When
    const {container, debug} = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DummyComponent/>
      </MockedProvider>
    )

    // Then
    expect(spy).toHaveBeenCalledTimes(0)
  })

  it("renders graphQL result without crashing", async () => {
    // Given

    const mocks = [
      {
        request: {query: mockQuery, variables},
        result: {
          data: {
            [typename]: {
              totalCount: 3,
              pageInfo: {
                endCursor: Symbol("endCursor"),
                hasNextPage: false
              },
              edges: [
                {
                  node: {id: 1}
                },
                {
                  node: {id: 2}
                },
                {
                  node: {id: 3}
                }
              ]
            }
          }
        }
      }
    ]
    const exp = [{id: 1}, {id: 2}, {id: 3}]

    function DummyComponent() {
      const queryProps = useQuery(mockQuery, {variables})
      return (
        <GraphQLInfiniteScroll typename={typename} {...queryProps}>
          {({records, total}) => {
            expect(records).toStrictEqual(exp)
            expect(total).toBe(3)
            return <div/>
          }}
        </GraphQLInfiniteScroll>
      )
    }

    // When
    const {container, debug} = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DummyComponent/>
      </MockedProvider>
    )
    await waitForDomChange({container})
  })

  it("renders error message without crashing when error occurred", async () => {
    // Given
    const mocks = [
      {
        request: {query: mockQuery, variables},
        error: new Error("Some Error")
      }
    ]

    // When
    const {container, debug} = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DummyComponent/>
      </MockedProvider>
    )
    await waitForDomChange({container})

    // Then
    expect(container).toMatchSnapshot()
  })

  it("renders nothing without crashing when query target not found", async (done) => {
    // Given
    const mocks = [
      {
        request: {query: mockQuery, variables},
        result: {
          data: {
            [typename]: undefined
          }
        }
      }
    ]

    // When
    const {container, debug} = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <DummyComponent/>
      </MockedProvider>
    )
    try {
      await waitForDomChange({container, timeout: 0.1})
    } catch {
      // Then
      done()
      expect(container.children).toHaveLength(0)
    }

  })

  it("renders nothing without crashing if typename is wrong", async (done) => {
    // Given
    const mocks = [
      {
        request: {query: mockQuery, variables},
        result: {
          data: {
            wrongtype: {
              totalCount: 3,
              pageInfo: {
                endCursor: Symbol("endCursor"),
                hasNextPage: false
              },
              edges: [
                {
                  node: {id: 1}
                },
                {
                  node: {id: 2}
                },
                {
                  node: {id: 3}
                }
              ]
            }
          }
        }
      }
    ]

    function WrongTypenameComponent() {
      const queryProps = useQuery(mockQuery, {variables})
      return (
        <GraphQLInfiniteScroll typename={"anotherTypeName"} {...queryProps}>
          {({data}) => JSON.stringify(data)}
        </GraphQLInfiniteScroll>
      )
    }

    // When
    const {container, debug} = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <WrongTypenameComponent/>
      </MockedProvider>
    )
    try {
      await waitForDomChange({container, timeout:0.1})
    } catch {
      // Then
      done()
      expect(container.children).toHaveLength(0)
    }

  })
})

describe("graphqlConcatResultNodes", () => {
  const typename = "dummyType"

  it("concatinates prev and next graphql node data", () => {
    const prev = {
      [typename]: {
        totalCount: 42,
        pageInfo: {
          hasNextPage: true,
          endCursor: Symbol("endCursor1")
        },
        edges: [
          {
            node: {id: 1}
          },
          {
            node: {id: 2}
          }
        ]
      }
    }
    const next = {
      [typename]: {
        totalCount: 42,
        pageInfo: {
          hasNextPage: true,
          endCursor: Symbol("endCursor2")
        },
        edges: [
          {
            node: {id: 3}
          },
          {
            node: {id: 4}
          }
        ]
      }
    }
    const exp = {
      [typename]: {
        totalCount: 42,
        pageInfo: {
          hasNextPage: true,
          endCursor: Symbol("endCursor2")
        },
        edges: [
          {
            node: {id: 1}
          },
          {
            node: {id: 2}
          },
          {
            node: {id: 3}
          },
          {
            node: {id: 4}
          }
        ]
      }
    }

    const res = graphqlConcatResultNodes(typename, prev, {fetchMoreResult: next})
    // INFO object同士の比較だとerrorが発生するため、serializeしておく
    expect(JSON.stringify(res)).toEqual(JSON.stringify(exp))
  })

  it("updates only pageInfo when fetchMoreResult has no edge", () => {
    const prev = {
      [typename]: {
        totalCount: 42,
        pageInfo: {
          hasNextPage: true,
          endCursor: Symbol("endCursor1")
        },
        edges: [
          {
            node: {id: 1}
          },
          {
            node: {id: 2}
          }
        ]
      }
    }
    const next = {
      [typename]: {
        totalCount: 42,
        pageInfo: {
          hasNextPage: false,
          endCursor: Symbol("endCursor2")
        },
        edges: []
      }
    }
    const exp = {
      [typename]: {
        totalCount: 42,
        pageInfo: {
          hasNextPage: false,
          endCursor: Symbol("endCursor2")
        },
        edges: [
          {
            node: {id: 1}
          },
          {
            node: {id: 2}
          }
        ]
      }
    }

    const res = graphqlConcatResultNodes(typename, prev, {fetchMoreResult: next})
    // INFO object同士の比較だとerrorが発生するため、serializeしておく
    expect(JSON.stringify(res)).toEqual(JSON.stringify(exp))
  })

  it("returns prev object as its original state when next is not defined", () => {
    const prev = Symbol("prev")

    const res = graphqlConcatResultNodes(typename, prev, undefined)
    expect(res).toBe(prev)
  })

  it("returns prev object as its original state when next.fetchMoreResult is not defined", () => {
    const prev = Symbol("prev")

    const res = graphqlConcatResultNodes(typename, prev, {fetchMoreResult: undefined})
    expect(res).toBe(prev)
  })
})
