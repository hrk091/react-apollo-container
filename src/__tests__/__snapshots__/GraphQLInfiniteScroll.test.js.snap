// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphQLInfiniteScroll renders error message without crashing when error occurred 1`] = `
<div>
  <h3>
    データ取得エラー
  </h3>
  <h5>
    エラーが発生しました。ページをリロードしてください。
  </h5>
  <h5>
    事象が改善しない場合は、管理者にお問い合わせください。
  </h5>
</div>
`;
